src/vidpf.rs

```diff
@@ -0,0 +1,676 @@
+            let w_i_1;
+            (s_i[0], w_i_0) = self.convert(&s_tilde_i_0, parameter);
+            (s_i[1], w_i_1) = self.convert(&s_tilde_i_1, parameter);
+
+            let mut w_cw = (beta - w_i_0.borrow()) + w_i_1;
+            w_cw.conditional_negate(t_i[1]);
+
+            let cw_i = CorrectionWord {
+                s_cw,
+                t_cw_l,
+                t_cw_r,
+                w_cw,
+            };
+            cw.push(cw_i);
+
+            let pi_tilde_0 = &Self::hash_one(alpha, level, &s_i[0]);
+            let pi_tilde_1 = &Self::hash_one(alpha, level, &s_i[1]);
+            let cs_i = xor_proof(pi_tilde_0, pi_tilde_1);
+            cs.push(cs_i);
+        }
+
+        Ok((PublicShare { cw, cs }, k0, k1))
+    }
+
+    /// [Vidpf::eval_next] queries one bit `(alpha_i)` in the evaluation tree at the level
+    /// corresponding to the current state `(s_i,t_i,cw_i)`. This evaluation updates
+    /// the state `(s_i,t_i)` and returns a new secret-shared value `(y)`.
+    #[must_use]
+    pub fn eval_next(
+        &self,
+        b: ServerId,
+        alpha_i: ControlBit,
+        s_i: &mut Seed,
+        t_i: &mut ControlBit,
+        cw_i: &CorrectionWord<V>,
+        parameter: &V::ValueParameter,
+    ) -> V {
+        let CorrectionWord {
+            s_cw,
+            t_cw_l,
+            t_cw_r,
+            w_cw,
+        } = cw_i;
+
+        let seq_tilde = &self.prg(s_i);
+
+        let sl = conditional_xor_seeds(&seq_tilde.0, s_cw, *t_i);
+        let sr = conditional_xor_seeds(&seq_tilde.2, s_cw, *t_i);
+        let tl = seq_tilde.1 ^ (*t_i & *t_cw_l);
+        let tr = seq_tilde.3 ^ (*t_i & *t_cw_r);
+
+        let s_tilde_i = conditional_select_seed(alpha_i, &[sl, sr]);
+        *t_i = ControlBit::conditional_select(&tl, &tr, alpha_i);
+
+        let w_i;
+        (*s_i, w_i) = self.convert(&s_tilde_i, parameter);
+
+        let mut y_i = V::conditional_select(&w_i, &(w_i.borrow() + w_cw), *t_i);
+        y_i.conditional_negate(Choice::from(b as u8));
+
+        y_i
+    }
+
+    /// [Vidpf::proof_next] generates a new [Proof] at the current level `(alpha,level)` from the
+    /// current state `(s_i,t_i,cs_i)` and current proof `(pi)`.
+    #[must_use]
+    pub fn proof_next(
+        &self,
+        pi: &Proof,
+        alpha: &VidpfInput,
+        level: usize,
+        s_i: &Seed,
+        t_i: ControlBit,
+        cs_i: &Proof,
+    ) -> Proof {
+        let pi_tilde = &Self::hash_one(alpha, level, s_i);
+        let h2_input = xor_proof(pi, &conditional_xor_proof(pi_tilde, cs_i, t_i));
+        xor_proof(pi, &Self::hash_two(&h2_input))
+    }
+
+    /// [Vidpf::eval] queries the evaluation tree with `alpha` input and produces a share of the output value.
+    ///
+    /// # Panics
+    /// Panics when the public key does not contain as many correction words as the length
+    /// of the `alpha` input in bits.
+    #[must_use]
+    pub fn eval(
+        &self,
+        alpha: &VidpfInput,
```

Comment:
We're virtually always going to need a sequence of candidate prefixes, not just one.
```suggestion
        prefixes: impl Iterator<Item = VidpfInput>,
```

Reply:
I think this api will be useful later, when using cache between invocations
so for now, let's process one input each time